### PR TITLE
[service] Fix memory leaks and enable goleak check in tests

### PR DESCRIPTION
**Pull request by crobert-1**
**Description:** <Describe what has changed.>

This change adds `goleak` to check for memory leaks. Originally there were 3 failing tests in the `service` package, so I'll describe changes in relation to resolving each test's failing goleak check.

1. `TestServiceTelemetryRestart`: Simplest fix, close the response body to make sure goroutines aren't leaked by reopening a server on the same port. This was just a test issue
2. `TestTelemetryInit.UseOTelWithSDKConfiguration`: The [meter provider](https://github.com/open-telemetry/opentelemetry-collector/blob/fb3ed1b0d65b91e49209a7e60d40ef4b607c6b10/service/telemetry.go#L57-L58) was being started in the initialization process ([metrics reference](https://github.com/open-telemetry/opentelemetry-collector/blob/fb3ed1b0d65b91e49209a7e60d40ef4b607c6b10/service/internal/proctelemetry/config.go#L135)), but never shutdown. The type originally being used (`meter.MetricProvider`) was the base interface which didn't provide a `Shutdown` method. I changed this to use the `sdk` interfaces that provide the required `Shutdown` method. The actual functionality of starting the providers was already using and returning the `sdk` interface, so the actual underlying type remains the same. Since `mp` is a private member and `sdkmetric` and implement the original type, I don't believe changing the type is a breaking change.
3. `TestServiceTelemetryCleanupOnError`: This test starts a server using a sub-goroutine, cancels it, starts again in a subroutine, and cancels again in the main goroutine. This test showed the racing behavior of the subroutine running [`server.ListenAndServe`](https://github.com/open-telemetry/opentelemetry-collector/blob/fb3ed1b0d65b91e49209a7e60d40ef4b607c6b10/service/internal/proctelemetry/config.go#L148) and the main goroutine's functionality of [calling close](https://github.com/open-telemetry/opentelemetry-collector/blob/fb3ed1b0d65b91e49209a7e60d40ef4b607c6b10/service/telemetry.go#L219) and then starting the server again [right away](https://github.com/open-telemetry/opentelemetry-collector/blob/fb3ed1b0d65b91e49209a7e60d40ef4b607c6b10/service/service_test.go#L244). The solution here is to add a `sync.WaitGroup` variable that can properly block until all servers are closed before returning from `shutdown`. This will allow us to ensure it's safe to proceed knowing the ports are free, and server is fully closed.

The first test fix was just a test issue, but 2 and 3 were real bugs. I realize it's a bit hard to read with them all together, but I assumed adding PR dependency notes would be more complicated.

**Link to tracking Issue:** <Issue number if applicable>
#9165

**Testing:** <Describe what testing was performed and which tests were added.>
All tests are passing as well as goleak check.

### CODE REVIEW

File: service/telemetry.go [hunk 1]
```diff
@@ -54,19 +51,21 @@ var (
...
 	disableHighCardinality bool
 	extendedConfig         bool
 }
 
 func newColTelemetry(useOtel bool, disableHighCardinality bool, extendedConfig bool) *telemetryInitializer {
 	return &telemetryInitializer{
-		mp:                     noopmetric.NewMeterProvider(),
-		tp:                     nooptrace.NewTracerProvider(),
+		mp:                     sdkmetric.NewMeterProvider(),
+		tp:                     sdktrace.NewTracerProvider(),
```
Comment (bogdandrutu) [inline service/telemetry.go:68]: Why this change?

Comment (crobert-1) [reply service/telemetry.go:68]: I wrote the reasoning here in the description, under point number 2. Let me know if any more information would be helpful.

File: service/telemetry.go [hunk 2]
```diff
@@ -214,11 +213,16 @@ func (tel *telemetryInitializer) shutdown() error {
...
 	var errs error
+	errs = multierr.Append(errs, tel.tp.Shutdown(context.Background()))
+	errs = multierr.Append(errs, tel.mp.Shutdown(context.Background()))
+
 	for _, server := range tel.servers {
 		if server != nil {
 			errs = multierr.Append(errs, server.Close())
 		}
 	}
+	tel.serverWG.Wait()
```
[the remark this answers is not in the document]
Comment (crobert-1) [reply service/telemetry.go:224]: Hey @michalpristas, good question! From my understanding, I don't believe this will ever block on shutdown.

This `Wait` call will only be hit once [`Close`](https://pkg.go.dev/net/http#Server.Close) has been called on every relevant server. Relevant servers in this case (`tel.servers`) are all started in the goroutine that you've referenced in `InitPrometheusServer`, so every started server increments the wg's counter. 

I'll paste the concerning `asyncErrorChannel` reference in question:
```
		if serveErr := server.ListenAndServe(); serveErr != nil && !errors.Is(serveErr, http.ErrServerClosed) {
			asyncErrorChannel <- serveErr
		}
```

The [`ListenAndServe` method](https://pkg.go.dev/net/http#Server.ListenAndServe) blocks until an error is hit, or `Shutdown`/`Close` is called. In the case of an error being returned by the server, a message is sent to the `asyncErrorChannel` channel and the goroutine will exit calling `wg.Done`. For the case of `Shutdown` or `Close` being called, `ListenAndServe` documentation states that the error is returned **after** the server is closed. No errors can be sent from the server once the server is closed, they'd all be sent beforehand. From these cases, I don't believe we have to worry about `Wait` blocking in `shutdown`.

Side note, this wait group was actually added because of the exact concern you're raising, if I'm following right. The test [`TestServiceTelemetryCleanupOnError`](https://github.com/open-telemetry/opentelemetry-collector/blob/8131d31601b886ad90b2cc8ae2e83ea96a9e2ca6/service/service_test.go#L236) was leaking a goroutine intermittently. The test would actually attempt to start a server on the same port twice in a row, and if it hit it fast enough, the second bind request would be made before `Close` was complete. This would cause an error to be sent to the `asyncErrorChannel` after the test was complete, causing the situation you've referenced.

If `shutdown` is called, it should be able to handle the running goroutines for any situation. Let me know if I misunderstood though.

[the remark this answers is not in the document]
Comment (michalpristas) [reply service/telemetry.go:175]: the problem i'm seeing is very sensitive to timing and we may not see it most of the time.
but i still think that replacing
```
asyncErrorChannel <- serveErr
```

with 
```
select{
case asyncErrorChannel <- serveErr:
default:
}
```

would not be that intrusive and could prevent leaks/locks on shutdown

[the remark this answers is not in the document]
Comment (crobert-1) [reply service/telemetry.go:175]: My apologies, but I'm not seeing how this will prevent a leak or stop the `tel.serverWG.Wait()` call from blocking. Can you share more detail on how the timing of the channel send operation `asyncErrorChannel <- serveErr` could cause a problem here? My understanding is that within this function there's no concurrency. The operation sending `serveErr` to `asyncErrorChannel` blocks until it's complete. If `asyncErrorChannel` is already closed it will panic. Is there some possible situation where the send operation will block forever, and that would cause the blocking `Wait` call you've referenced?

I agree your proposed `select` statement isn't a huge impacting call, but I want to avoid introducing timing delays as a solution as much as possible.

[the remark this answers is not in the document]
Comment (crobert-1) [reply service/telemetry.go:132]: I've done some more research and seen that this is a pretty technique in go, so I've added a `1s` timeout. I'd still be interested to hear though how it could block in this situation.

**Comment by mx-psi** [comment]
@crobert-1 What's the status of this PR? Does it need more reviews? Do you need to fix something? (No rush, just trying to clean up our backlog)

**Comment by crobert-1** [comment]
> @crobert-1 What's the status of this PR? Does it need more reviews? Do you need to fix something? (No rush, just trying to clean up our backlog)

Just waiting for reviews 👍

**Review by mx-psi — APPROVED** [review]

**Review by mx-psi — COMMENTED** [review]
Can you fix the merge conflicts? I can merge after that

**Comment by crobert-1** [comment]
> Can you fix the merge conflicts? I can merge after that

Should be good now, thanks @mx-psi!
